hardware_interface/include/hardware_interface/hardware_component_interface.hpp

```diff
@@ -703,30 +703,117 @@ class HardwareComponentInterface : public rclcpp_lifecycle::node_interfaces::Lif
+          "The requested state interface not found: '{}' in hardware component: '{}'.",
+          interface_name, info_.name));
+    }
+    return it->second;
+  }
+
+  /// Set the value of a state interface.
+  /**
+   * \tparam T The type of the value to be stored.
+   * \param interface_handle The shared pointer to the state interface to access.
+   * \param value The value to store.
+   * \param ensure_set If true, the method ensures that the value is set successfully
+   * \note This method is not real-time safe, when used with ensure_set = true.
+   * \throws std::runtime_error This method throws a runtime error if it cannot
+   * access the state interface.
+   * \return True if the value was set successfully, false otherwise.
+   */
+  template <typename T>
+  bool set_state(
+    const StateInterface::SharedPtr & interface_handle, const T & value, bool ensure_set = true)
+  {
+    if (!interface_handle)
+    {
+      throw std::runtime_error(
+        fmt::format(
+          "State interface handle is null in hardware component: {}, while calling set_state "
+          "method. This should not happen.",
+          info_.name));
+    }
+    if (ensure_set)
+    {
+      std::unique_lock<std::shared_mutex> lock(interface_handle->get_mutex());
+      return interface_handle->set_value(lock, value);
+    }
+    else
+    {
+      return interface_handle->set_value(value);
+    }
+  }
+
   /// Set the value of a state interface.
   /**
    * \tparam T The type of the value to be stored.
    * \param[in] interface_name The name of the state interface to access.
    * \param[in] value The value to store.
+   * \param ensure_set If true, the method ensures that the value is set successfully
+   * \note This method is not real-time safe, when used with ensure_set = true.
    * \throws std::runtime_error This method throws a runtime error if it cannot
    * access the state interface.
    */
   template <typename T>
-  void set_state(const std::string & interface_name, const T & value)
+  void set_state(const std::string & interface_name, const T & value, bool ensure_set = true)
   {
-    auto it = hardware_states_.find(interface_name);
-    if (it == hardware_states_.end())
+    std::ignore = set_state(get_state_interface_handle(interface_name), value, ensure_set);
+  }
+
+  /**
+   * \tparam T The type of the value to be retrieved.
+   * \param[in] interface_handle The shared pointer to the state interface to access.
+   * \param[out] state The variable to store the retrieved value.
+   * \param[in] ensure_get If true, the method ensures that the value is retrieved successfully
+   * \note This method is not real-time safe, when used with ensure_get = true.
+   * \throws std::runtime_error This method throws a runtime error if it cannot
+   * access the state interface or its stored value, when ensure_get is true.
+   * \return True if the value was retrieved successfully, false otherwise.
+   */
+  template <typename T>
+  bool get_state(
```

Comment:
For get_state basically the same reasoning applies as for set_state

Reply:
I modified a bit